src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -428,83 +450,74 @@ private async Task RefreshIndividualKeyValues(CancellationToken cancellationToke
-                    {
-                        KeyValueChange keyValueChange = default;
-                        await TracingUtils.CallWithRequestTracing(_requestTracingEnabled, RequestType.Watch, _requestTracingOptions,
-                            async () => keyValueChange = await _client.GetKeyValueChange(watchedKv, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
+                    KeyValueChange keyValueChange = default;
+                    await TracingUtils.CallWithRequestTracing(_requestTracingEnabled, RequestType.Watch, _requestTracingOptions,
+                        async () => keyValueChange = await client.GetKeyValueChange(watchedKv, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
 
-                        success = true;
 
-                        // Check if a change has been detected in the key-value registered for refresh
-                        if (keyValueChange.ChangeType != KeyValueChangeType.None)
+                    // Check if a change has been detected in the key-value registered for refresh
+                    if (keyValueChange.ChangeType != KeyValueChangeType.None)
+                    {
+                        if (changeWatcher.RefreshAll)
                         {
-                            if (changeWatcher.RefreshAll)
-                            {
-                                shouldRefreshAll = true;
-                                break;
-                            }
-
-                            if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
-                            {
-                                _watchedSettings.Remove(watchedKeyLabel);
-                            }
-                            else if (keyValueChange.ChangeType == KeyValueChangeType.Modified)
-                            {
-                                _watchedSettings[watchedKeyLabel] = keyValueChange.Current;
-                            }
-
-                            hasChanged = true;
-                            ProcessChanges(Enumerable.Repeat(keyValueChange, 1));
+                            shouldRefreshAll = true;
+                            break;
                         }
-                    }
-                    else
-                    {
-                        // Load the key-value in case the previous load attempts had failed
-                        var options = new SettingSelector { LabelFilter = watchedLabel };
 
-                        try
+                        if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
                         {
-                            await CallWithRequestTracing(async () => watchedKv = await _client.GetConfigurationSettingAsync(watchedKey, watchedLabel, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
+                            _watchedSettings.Remove(watchedKeyLabel);

```

Comment:
With this change, we are treating the entire refresh operation as one single workflow, i.e., if we detect any change using one client, we must complete the entire config refresh operation with that client. If at any point during the refresh workflow, we encounter an error that is eligible for failover, we must restart the entire refresh workflow with a new client. This means that we need to be careful not to change the state of any persisted data during the refresh workflow, because at any point we could have to start over. This includes: 
- cache expiration time for each change watcher 
- cached values of sentinel keys in `_watchedSettings`
- cached key-values in `_applicationSettings`

If we update any of these as soon as we get a successful response from server, we are losing their original state from before the refresh operation. We need the original data to make important decisions during refresh and we cannot get it back if we need to failover and start again.

Reply:
Do we have this problem with existing code-base between two back-to-back refresh calls?

Reply:
In the current code, when there are multiple watchers, we may refresh one watcher with client A and refresh another watcher with client B. We can accept this, but this is different than what we discussed.